dpe/src/validation.rs

```diff
@@ -0,0 +1,505 @@
+// Licensed under the Apache-2.0 license.
+
+use crate::{
+    context::{Context, ContextState, ContextType},
+    dpe_instance::flags_iter,
+    response::DpeErrorCode,
+    tci::TciNodeData,
+    DpeInstance, MAX_HANDLES,
+};
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(u32)]
+/// It is possible that there are multiple issues with the DPE state. At most one will be found.
+/// There is no priority on which error will be found first if there are multiple.
+pub enum ValidationError {
+    MultipleNormalConnectedComponents = 0x0,
+    CyclesInTree = 0x1,
+    InactiveContextInvalidParent = 0x2,
+    InactiveContextWithChildren = 0x3,
+    BadContextState = 0x4,
+    BadContextType = 0x5,
+    InactiveContextWithMeasurement = 0x6,
+    MixedContextLocality = 0x7,
+    MultipleDefaultContexts = 0x8,
+    SimulationNotSupported = 0x9,
+    ParentDoesNotExist = 0xA,
+    InternalDiceNotSupported = 0xB,
+    InternalInfoNotSupported = 0xC,
+    ChildDoesNotExist = 0xD,
+    InactiveContextWithFlagSet = 0xE,
+    LocalityMismatch = 0xF,
+    DanglingRetiredContext = 0x10,
+    MixedContextTypeConnectedComponents = 0x11,
+    ChildWithMultipleParents = 0x12,
+    ParentChildLinksCorrupted = 0x13,
+}
+
+impl ValidationError {
+    pub fn discriminant(&self) -> u16 {
+        // SAFETY: Because `Self` is marked `repr(u16)`, its layout is a `repr(C)` `union`
+        // between `repr(C)` structs, each of which has the `u16` discriminant as its first
+        // field, so we can read the discriminant without offsetting the pointer.
+        unsafe { *<*const _>::from(self).cast::<u16>() }
+    }
+}
+
+pub struct DpeValidator<'a> {
+    pub dpe: &'a mut DpeInstance,
+}
+
+impl<'a> DpeValidator<'a> {
+    /// Validates that the shape of the DPE instance is well-formed and that
+    /// there is no illegal state present within the DPE.
+    pub fn validate_dpe(&self) -> Result<(), DpeErrorCode> {
+        self.validate_dpe_state()
+            .map_err(DpeErrorCode::Validation)?;
+        self.validate_context_forest()
+            .map_err(DpeErrorCode::Validation)
+    }
+
+    /// Returns an error if there is any illegal state or inconsistencies
+    /// present within the DPE instance.
+    fn validate_dpe_state(&self) -> Result<(), ValidationError> {
+        for i in 0..MAX_HANDLES {
+            let context = &self.dpe.contexts[i];
+
+            self.check_support(context)?;
+
+            match context.state {
+                ContextState::Inactive => {
+                    self.validate_inactive_context(context)?;
+                }
+                ContextState::Active => {
+                    self.check_children_and_parent(i)?;
```

Comment:
We should check that if there is an active non-simulation context `self.dpe.has_initialized` is true, otherwise we could have a second initialized tree by accident.

Reply:
Now checking that if there is an active, normal ctx, has_initialized must be set to true.